frappe/oauth.py

```diff
@@ -164,28 +154,72 @@ def validate_code(self, client_id, code, client, request, *args, **kwargs):
 		# Validate the code belongs to the client. Add associated scopes,
 		# state and user to request.scopes and request.user.
 
-		validcodes = frappe.get_all("OAuth Authorization Code", filters={"client": client_id, "validity": "Valid"})
+		validcodes = frappe.get_all(
+			"OAuth Authorization Code",
+			filters={"client": client_id, "validity": "Valid"},
+		)
 
 		checkcodes = []
 		for vcode in validcodes:
 			checkcodes.append(vcode["name"])
 
 		if code in checkcodes:
-			request.scopes = frappe.db.get_value("OAuth Authorization Code", code, 'scopes').split(get_url_delimiter())
-			request.user = frappe.db.get_value("OAuth Authorization Code", code, 'user')
+			request.scopes = frappe.db.get_value(
+				"OAuth Authorization Code", code, "scopes"
+			).split(get_url_delimiter())
+			request.user = frappe.db.get_value("OAuth Authorization Code", code, "user")
+			code_challenge_method = frappe.db.get_value(
+				"OAuth Authorization Code", code, "code_challenge_method"
+			)
+			code_challenge = frappe.db.get_value(
+				"OAuth Authorization Code", code, "code_challenge"
+			)
+
+			if code_challenge and not request.code_verifier:
+				if frappe.db.exists("OAuth Authorization Code", code):
+					frappe.delete_doc(
+						"OAuth Authorization Code", code, ignore_permissions=True
+					)
+					frappe.db.commit()
+				return False
+
+			if code_challenge_method == "s256":
+				m = hashlib.sha256()
+				m.update(bytes(request.code_verifier, "utf-8"))
+				code_verifier = base64.b64encode(m.digest()).decode("utf-8")
+				code_verifier = re.sub(r"\+", "-", code_verifier)
+				code_verifier = re.sub(r"\/", "_", code_verifier)
+				code_verifier = re.sub(r"=", "", code_verifier)
```

Comment:
The [pkce](https://github.com/RomeoDespres/pkce/blob/master/pkce/__init__.py#L94) library encodes with  `ascii` instead of `utf-8`. Would that help avoid needing the substitutions, or are we looking at different scenarios?

Reply:
No need for library just for this one functionality.

Reply:
No no, I just meant the encoding-decoding parts on lines 188-189

Reply:
```
>>> import hashlib, base64
>>> m = hashlib.sha256()
>>> m.update(bytes("420", "ascii"))
>>> code_verifier = base64.b64encode(m.digest()).decode("ascii")
>>> code_verifier
'21XaP8MJjpxCMRxgEzBP82sZ73PRLqkyBUta1R309J0='

# expected 21XaP8MJjpxCMRxgEzBP82sZ73PRLqkyBUta1R309J0
```

It seems we still need to replace those characters.

Reply:
Ah, okay. They seemed to be removing just the last `=` in the string. If we need more substitutions, then this is fine.